### PR TITLE
Eq 374 serialiser

### PR DESCRIPTION
**Pull request by weapdiv-david**
### Changes

Adds the MCI Schema as python code, along with to_json() methods to generate json from model.
Also overrides **eq** and **ne** methods to check for equivalence rather than strict equality.
### How to test

1) Check out the branch
2) Run the unit tests
3) Verify they all pass
### Who can test

Anyone except @weapdiv-david

### CODE REVIEW

File: app/model/block.py [hunk 1]
```diff
@@ -10,11 +10,73 @@ def __init__(self):
...
+        if self.validation is not None:
+            for validation in self.validation:
+                json_dict['validation'].append(validation.to_json())
+
+        if self.display is not None:
+            json_dict['display'] = self.display.to_json()
+
+        return json_dict
+
+    def __eq__(self, other):
```
Comment (collisdigital) [inline app/model/block.py:43]: You've got to be careful with this, anywhere you'd normally use a == (or code that performs the same comparison) is going to call this code and not compare the object pointers; which may not be obvious to the caller and give strange behaviour if not expected, if you do stick with it then you probably need pretty good unit tests around the operator to ensure it is working correctly (i.e. verifies each change in property results in a not equal outcome etc.). This seems like a lot of code to create and maintain just to test the json stuff; do you really want to do/keep these methods, another approach would be to run this through the to_json method, dump the resulting JSON text and paste that text into the unit test to verify that the code generates the expected json text or manually in the unit test check the properties of the objects etc. Worth discussing with @dhilton @LJBabbage ?

File: app/surveys/mci/mci_0205.py [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+s2.description = "<p>- You should enter figures for the reporting period stated above<br>- You should round your figures to the nearest (£) pound</p><h4>Include</h4><ul><li>VAT</li><li>Internet Sales</li></ul>"  # NOQA
+
+block.add_section(s2)
+questionnaire.register(s2)
+
+q2 = Question()
+q2.id = "eedb9f9f-2f2e-41f4-9186-ba7110d9e6a4"
+q2.title = ""
+q2.description = ""
+q2.type = "Currency"
```
Comment (collisdigital) [inline app/surveys/mci/mci_0205.py:112]: In the latest agreed model all Questions would be generic/general and only specialised question types like DateRange would have a different type. Maybe not to do now but at some point (sooner rather than later) we should make the switch to this approach.

File: app/surveys/mci/mci_0205.py [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+
+s2.add_question(q2)
+questionnaire.register(q2)
+
+q2r = Response()
```
Comment (collisdigital) [inline app/surveys/mci/mci_0205.py:117]: We discussed the possibility of the different Response types having their own classes (e.g. CurrencyResponse) etc. Another one of those when would be the right time to make that jump?

File: app/surveys/mci/mci_0205.py [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+q3r = Response()
+q3r.id = "fee0b9fe-4c3a-4c14-9611-4fa9e2e9578a"
+q3r.code = "23"
+q3r.label = "What was the value of the business's total sales of alcohol, confectionery and tobacco?"
+q3r.guidance = "<div> <h4>Include</h4> <ul> <li>alcoholic drink</li> <li>chocolate and sugar confectionery</li> <li>tobacco and smokers’ requisites</li></ul> </div>"  # NOQA
+q3r.type = "Currency"
+q3r.options = []
+q3r.mandatory = False
+q3r.messages = {
+    "NOT_INTEGER": "Please only enter whole numbers into the field.",
```
Comment (collisdigital) [inline app/surveys/mci/mci_0205.py]: Are these keys stored anywhere, I mean "NOT_INTEGER" that could be used instead (enum?) instead of pasting in the strings directly?

File: app/model/block.py [hunk 5]
```diff
@@ -10,11 +10,73 @@ def __init__(self):
...
+        self.questionnaire = None
         self.templatable_properties = []
         self.display = Display()
 
     def add_section(self, section):
         if section not in self.sections:
             self.sections.append(section)
             section.container = self
+
+    def to_json(self):
```
Comment (collisdigital) [inline app/model/block.py:22]: When we did the design review with Warren it was agreed to also implement a from_json method that did the reverse operation so you could do a round trip to/from JSON, is this still in scope?

File: app/model/block.py (hunk 1 above)
Comment (weapdiv-david) [inline app/model/block.py:43]: Definitely worth discussing with the larger team for the reasons you mentioned.  I had to change a number of the unittests to ensure that some objects being created were no longer equivalent, where the tests had simply required that they were not the same object.

I'm open to alternative approaches

File: app/surveys/mci/mci_0205.py (hunk 2 above)
Comment (weapdiv-david) [inline app/surveys/mci/mci_0205.py:112]: Agreed this needs to happen soon, but was left out of this in order to limit the scope.  Otherwise, the parser needs to change, etc...

File: app/surveys/mci/mci_0205.py (hunk 3 above)
Comment (weapdiv-david) [inline app/surveys/mci/mci_0205.py:117]: Similar to above, this was left out _for now_ in order to limit the scope and minimise the impact of other parts of the system.  It should probably be done sooner rather than later though, but perhaps a different PR

File: app/surveys/mci/mci_0205.py (hunk 4 above)
Comment (weapdiv-david) [inline app/surveys/mci/mci_0205.py]: My Bad!  They are stored as constants... I'll refactor them now

File: app/model/block.py (hunk 5 above)
Comment (weapdiv-david) [inline app/model/block.py:22]: Could still be in scope... I'm not sure.  Currently the to_json() method works to output json in the format expected by the parser, so the round trip can be done through that as seen in the unit-tests.

It's not quite as clean as having a from_json() method, but while so much of this is up for grabs, I'm not sure how much value it adds at this stage to the implement that method.

Any thoughts from @dhilton, @LJBabbage ?

File: app/model/block.py (hunk 1 above)
Comment (dhilton) [inline app/model/block.py:43]: Reading around the subject, I still think this is the right approach for this instance. We also need to add `__ne__` for completeness, but given we're writing a serialisation / deserialisation from JSON we need to be able to compare the objects generated correctly. 

I'm trying to think of an example where we'd compare two objects that would require us to check the internal python id of that object (which is how python does this check by default when comparing objects).

Comment (weapdiv-david) [inline app/model/block.py:43]: `__ne__` has been implemented for each of the model objects.

Comment (weapdiv-david) [inline app/model/block.py:43]: With regard to doing a comparision at json string level, that approach didn't work as the order in which the properties appeared in the json seemed to vary, making the string comparison impossible without resorting to lots and lots of fragile regexps.

Parsing the generated schema and doing an object model comparison seemed like the better way forward but comes with the caveats discussed above around equivalence vs equality

Comment (weapdiv-david) [inline app/model/block.py:43]: As for checking the internal id, the overridden `__eq__` method check that check first as a shortcut before comparing the equivalence of properties (which are also compared for equivalence).

The main issue to be aware of is with collections.  Two equivalent but separate objects may not be stored in a set, for example.  Also, if an object is already in a list, checking that a different but equivalent object is in the list will return true, even through they are different objects

Comment (collisdigital) [inline app/model/block.py:43]: If you stick with the **eq** approach also check **hash** e.g. https://docs.python.org/3/reference/datamodel.html#object.__hash__ and I believe (but could be wrong that **ne** by default calls **eq** and inverts the result (i.e. no need to explicitly add it). Also consider better unit tests to cover the operators.

File: app/model/block.py (hunk 5 above)
Comment (dhilton) [inline app/model/block.py:22]: I think right now unless you've this locally, let's cut that from this story and migrate to a new one.

**Comment by dhilton** [comment]
LGTM - 👍
